### PR TITLE
Changes made to colordetection.py

### PR DESCRIPTION
**Pull request by Heydok**
Made some changes to colordetection.py and added a new file color_params.py used for parameter settings

### CODE REVIEW

File: src/colordetection.py [hunk 1]
```diff
@@ -21,20 +22,24 @@ def get_color_name(self, hsv):
         """
         (h,s,v) = hsv
         #print((h,s,v))
-        if h < 15 and v < 100:
-            return 'red'
-        if h <= 10 and v > 100:
-            return 'orange'
-        elif h <= 30 and s <= 100:
+        if s <= color_params.sat_W and v >= color_params.val_W:
```
Comment (kkoomen) [inline src/colordetection.py:25]: I trust your logic more than mine, but are you 100% that you should use `<=` and `>=` ? Because my logic is `<` and `<`

**Comment by kkoomen** [comment]
Hi, thanks for your time. Really appreciate it. I'd love to get some help to improve color detection much more. Your version is quite an improvement already, but there are still colors it guessed wrong, see below: 

- The orange is detected as red
- Green is detected as white

image

**Comment by HaginCodes** [comment]
Hey I've been following notifications from this repo. I used to do color detection like this too when I was starting out until I looked into creating an algorithm to do it for me that helped me achieve almost perfect accuracy every time. 

I will explain this in brief and send you a couple links too.
The way my color sorting and detection algorithm works is that I use the centers as the "core" colors. 
1. I then create 6 groups, one for each center.
2. I then take all the colors on the cube and compare them against my "core" colors.
3. I find the euclidean distance between the current core color and the "nth" color and store the distance. 
4. I store the lowest 8 euclidean numbers closest to the current "core" color and define them as that core color. 
5. I remove these 9 colors from the pool of 54 and repeat the process until there are no more colors left to pick from.

Tips: I have all this code available in my solver linked in my repo
Code I use to find centers:  
https://github.com/HaginCodes/3x3x3-Rubiks-Cube-Solver/blob/3b95db9094e48e082b78a737874ff50f2d9ca316/src/AnalyzeFrame.java#L260

Code for finding the rest of the stickers:  
https://github.com/HaginCodes/3x3x3-Rubiks-Cube-Solver/blob/3b95db9094e48e082b78a737874ff50f2d9ca316/src/AnalyzeFrame.java#L309
(Note: this is called k-means because it's the algorithm I used to sort the colors. I had to write it by hand because I was using Java but I'm sure you can find a python library that has it and you can use it to do the exact same thing

Finding the distance between colors, when I was working on my color detection algorithm, initially, I was using HSV too. However, I switched from it because it wasn't reliable enough colorspace when it came to certain colors like red and orange and other close colors depending on the ambient lighting. Options you can look into are HSL, YCbCr, and LaB. I personally used LaB and I found it worked best for me and was reliable in all lighting conditions. 

For calculating color distance there's two options you can choose but this depends on the color space you use. When I was using LaB, I had two options. Euclidean distance (very quick and arithmetic calculation light and quick) and the other was CIEDE2000 https://en.wikipedia.org/wiki/Color_difference#CIEDE2000. They both work in similar ways, and "apparently" CIEDE2000 is more accurate but I didn't notice any difference when using one or the other after very numerous tests.

If you have any further questions just let me know and good luck on your color detection algorithm, this was definitely the most annoying and tedious part of the whole project but once you get it working reliably it's so rewarding. Best of luck!

**Comment by kkoomen** [comment]
@HaginCodes Many thanks for the answer! I will definitely read your implementation and try to do it myself. Getting quite accurate results every time is something I'm really trying to achieve. This project started out 5 years ago just because I was a programmer and I liked solving cubes, but I've always wanted to optimize this, but never knew how you. Thanks a lot man! If I have any questions, I'll ask you! ;)

File: src/colordetection.py (hunk 1 above)
Comment (Heydok) [reply src/colordetection.py:17]: I am fairly sure that it is correct yes, but will run some tests tonight.

**Comment by kkoomen** [comment]
@Heydok I will first use another version based on what @HaginCodes mentioned. If his version is more accurate using LaB then I'll close this issue.

**Comment by kkoomen** [comment]
This PR has no use anymore, since I've finished the rewrite to use the CIEDE2000 algorithm.

@HaginCodes Many thanks for the suggestion. I didn't use the kmeans since that wasn't needed at all.

**Pull request closed without merging**